subsys/spm/Kconfig

```diff
@@ -4,12 +4,15 @@
 # SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
 #
 
-menuconfig SPM
-	bool "Secure Partition Manager"
-	default n
+config SPM
+	bool "Use Secure Partition Manager"
+	default y if BOARD_NRF9160_PCA10090NS
```

Comment:
Hi, just wanted to add an explanation as to what is going on here.

A non-secure image MUST have a (direct or indirect) SPM child image. This is because pure nonsecure images don't boot, the HW starts in secure mode, and then eventually jumps to nonsecure mode.

What we are doing here, is expressing that by default, if you are a nonsecure image, then you are going to want to "use" or "have" a second secure SPM child image.

Intuitively this is not a valid assumption, because you could have multiple nonsecure images. But as of right now there is a limitation that prevents more than one nonsecure image. So the assumption is valid and the configuration makes sense.

Technically, it would be more accurate to express if "nonsecure" than if "nrf91 nonsecure". Do we have a Kconfig for that?

Reply:
Let me describe an example:
Assume I want to compile the hello-world sample, or any other Zephyr application sample for the nrf9160_pca10090ns board. I realize this will be a Non-Secure image, which will need some sort of a secure-booting firmware, but, at this point, I don't care. I only want to build my sample for the _ns board. I could assume by board already has some magic Secure image inside.

I don't see a reason why CONFIG_SPM should be "y" in that case. Usually, in Zephyr, CONFIG_"susbsystem" means, we are building a Zephyr firmware with "subsystem" support. E.g. CONFIG_BT, for Bluetooth.

But my hello-world firmware does not use, at all, the SPM subsystem, inside its functionality. Therefore, it shall not have CONFIG_SPM defined. That's the same as it shall not have CONFIG_IPV6, CONFIG_SHELL, CONFIG_USB, etc. defined.

Second point: a sub-system is never board-dependent. We select subsystems in our application entries (prj.conf); this is the standard way of doing it. A subsystem could also depend on other subsystems, of course.

Now, I also notice you changed the text to "Use Secure Partition Manager". What is the meaning of this? A help text would help here.

Conclusion: ideally, for the SPM subsystem, we need a CONFIG_SPM to be defined, if our image (i.e. what we are now building) requires it in its functionality.

If we claim we keep supporting the _old_ way of doing things, i.e. someone just compiling a NS image and flashing it to the device (i.e. what we are supporting UPSTREAM) we need to fix that.

Otherwise, we break compatibility with upstream and with the earlier versions. IMHO we need to still support standalone Secure or Non-Secure builds of applications in the nrf9160 DK.

Reply:
Minor edit: in theory a sub-system could be BOARD-dependent (e.g. designed specifically for a BOARD), but then we select the sub-system in the BOARD definition.

Reply:
I think I agree with @ioannisg. The assumption that when building any non-secure sample the developer/user also wants to build `spm` and flash the two together is not always valid. Generally, when a user builds a sample, they want to do just that. It is reasonable to assume that `spm` or `mcuboot` or something else already exists on the board and the user does not want flashing or building a _sample_ to mess with other software images e.g. the user wants to keep the existing version of `spm` for testing. It is also a valid point to say that if a sample does not use a subsystem, that subsystem should not be turned on by default.

Reply:
@SebastianBoe ?

Reply:
Hi, we will get back to you soon about this.

Reply:
The first time a user compiles and flashes an example, they expect it to work, and see expected output after performing `ninja flash`. With the current default values, this is the case.

Now that the SPM is in place, the user can opt out of building it by setting the appropriate kconfig option. The choice is to either point to a hex file, which is merged in, or simply skip building it completely. Alternatively one can set `CONFIG_SPM=n`, and leave it at that.

IMO usability is the only reason for having multi image building, and IMO it is not user friendly to require the users to still have to manually compile a separate application in order to get a non-secure application working.

@ioannisg  your point about selecting a sub-system from BOARD is a bit moot IMO, since, ideally the "non-secure" configuration should come from Kconfig not the BOARD. So we should treat this as a Kconfig option deciding what subsystem is included, since that is what it will be in the final solution.

Let me know if anything is unclear, or if there was some points I missed.

If you mean that the usability argument is not sufficient, then I propose you get consensus among the code owners and change it back :)

Reply:
The board-specific issue is now resolved: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/644

Reply:
Whether the default is default y or default n produces these two different user-error scenarios;

User enables non-secure, flashes his application, and sees that it doesn't boot in spite of the configuration system giving the all-clear.

User enables non-secure, flashes his application, and remarks that his previously flashed spm image was overwritten by a new (possibly identical) one.

IMHO it is more important to protect against the first user-error than the second one. The first results in a hard-to-debug runtime error, the second will usually result in a couple of seconds build-time overhead.

Essentially, we have to guess what the common user is going to expect from the system. I believe more users are going to expect that enabling non-secure will pull in all the non-secure dependencies, than we will have users expecting that their previously flashed SPM image will be retained when flashing a new application image. (Presumably, by the time they have a custom-configured SPM they have already flashed an application and seen that SPM is included in when flashing the application).

NB: What the default value is does not affect what the user can or cannot do. All use-cases are covered regardless.

Reply:
> It is also a valid point to say that if a sample does not use a subsystem, that subsystem should not be turned on by default.

In this case, the sample/user has enabled non-secure, which in turn has enabled SPM as a dependency.

We are not defaulting to enabling a feature that the user/sample has not asked for.

Reply:
So this looked to be a sort of false-alarm. You never wanted to enable SPM for non-secure builds. 

If you had responded somehow to this part of my original comment:

>Now, I also notice you changed the text to "Use Secure Partition Manager". What is the meaning of this? A help text would help here.

... I would have also noticed how you changed the meaning of config SPM.
Instead, we kept discussing on the wrong basis :)

Reply:
Resolving this one, since we have the discussion in #644.